### PR TITLE
sql: do not close stmt buffer of internal executor in errCallback

### LINKED ISSUE #79746: sql: v21.1.6: attempting to push into closed stmt buffer in the internal executor
This issue was autofiled by Sentry. It represents a crash or reported error on a live cluster with telemetry enabled.

Sentry link: https://sentry.io/organizations/cockroach-labs/issues/3181618283/?referrer=webhooks_plugin

Panic message:

>conn_io.go:401: ×: buffer is closed
--
*errutil.leafError: buffer is closed (1)
conn_io.go:401: *withstack.withStack (top exception)
*assert.withAssertionFailure
*errutil.withPrefix: × (2)
internal.go:649: *withstack.withStack (3)
(check the extra data payloads)

<details>
<summary>Stacktrace (expand for inline code snippets):</summary>

https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/internal.go#L648-L650 in pkg/sql.(*InternalExecutor).execInternal.func1
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/internal.go#L743-L745 in pkg/sql.(*InternalExecutor).execInternal
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/internal.go#L409-L411 in pkg/sql.(*InternalExecutor).queryInternalBuffered
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/internal.go#L467-L469 in pkg/sql.(*InternalExecutor).QueryRowExWithCols
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/internal.go#L453-L455 in pkg/sql.(*InternalExecutor).QueryRowEx
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/catalog/lease/lease.go#L404-L406 in pkg/sql/catalog/lease.CountLeases
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/catalog/descs/txn.go#L133-L135 in pkg/sql/catalog/descs.CheckTwoVersionInvariant
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/conn_executor_exec.go#L720-L722 in pkg/sql.(*connExecutor).checkDescriptorTwoVersionInvariant
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/conn_executor_exec.go#L769-L771 in pkg/sql.(*connExecutor).commitSQLTransactionInternal
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/conn_executor_exec.go#L748-L750 in pkg/sql.(*connExecutor).commitSQLTransaction
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/conn_executor_exec.go#L1477-L1479 in pkg/sql.(*connExecutor).handleAutoCommit
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/conn_executor_exec.go#L482-L484 in pkg/sql.(*connExecutor).execStmtInOpenState.func7
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/conn_executor_exec.go#L712-L714 in pkg/sql.(*connExecutor).execStmtInOpenState
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/conn_executor_exec.go#L122-L124 in pkg/sql.(*connExecutor).execStmt
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/conn_executor.go#L1526-L1528 in pkg/sql.(*connExecutor).execCmd.func1
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/conn_executor.go#L1528-L1530 in pkg/sql.(*connExecutor).execCmd
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/conn_executor.go#L1449-L1451 in pkg/sql.(*connExecutor).run
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/conn_executor.go#L483-L485 in pkg/sql.(*Server).ServeConn
https://github.com/cockroachdb/cockroach/blob/03b83e02502e8876ecb366659156ab00f7431a51/pkg/sql/pgwire/conn.go#L627-L629 in pkg/sql/pgwire.(*conn).processCommandsAsync.func1
/usr/local/go/src/runtime/asm_amd64.s#L1373-L1375 in runtime.goexit
</details>

```
pkg/sql/internal.go in pkg/sql.(*InternalExecutor).execInternal.func1 at line 649
pkg/sql/internal.go in pkg/sql.(*InternalExecutor).execInternal at line 744
pkg/sql/internal.go in pkg/sql.(*InternalExecutor).queryInternalBuffered at line 410
pkg/sql/internal.go in pkg/sql.(*InternalExecutor).QueryRowExWithCols at line 468
pkg/sql/internal.go in pkg/sql.(*InternalExecutor).QueryRowEx at line 454
pkg/sql/catalog/lease/lease.go in pkg/sql/catalog/lease.CountLeases at line 405
pkg/sql/catalog/descs/txn.go in pkg/sql/catalog/descs.CheckTwoVersionInvariant at line 134
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).checkDescriptorTwoVersionInvariant at line 721
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).commitSQLTransactionInternal at line 770
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).commitSQLTransaction at line 749
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).handleAutoCommit at line 1478
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).execStmtInOpenState.func7 at line 483
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).execStmtInOpenState at line 713
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).execStmt at line 123
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).execCmd.func1 at line 1527
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).execCmd at line 1529
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).run at line 1450
pkg/sql/conn_executor.go in pkg/sql.(*Server).ServeConn at line 484
pkg/sql/pgwire/conn.go in pkg/sql/pgwire.(*conn).processCommandsAsync.func1 at line 628
/usr/local/go/src/runtime/asm_amd64.s in runtime.goexit at line 1374
```

| Tag | Value |
|---|---|
| Cockroach Release | `v21.1.6` |
| Cockroach SHA: | 03b83e02502e8876ecb366659156ab00f7431a51 |
| Platform | linux amd64 |
| Distribution | CCL |
| Environment | development |
| Command | start-single-node |
| Go Version | ``|
| # of CPUs ||
| # of Goroutines ||

### PR DESCRIPTION
**Pull request by yuzefovich**
Previously, we would close the stmt buffer of the internal executor in
`errCallback`, "just to be safe" since it was assumed that the buffer is
already closed when the callback is executed. The callback runs whenever
`run()` loop of connExecutor exits with an error.

However, it is possible for the following sequence of events to happen:
- The new goroutine is spun up for the internal executor before any
commands are pushed into the stmt buffer.
- The context is canceled before the new goroutine blocks waiting for
the command to execute, i.e. `run()` loop is exited before any commands
are executed.
- The `errCallback` with the context cancellation error is evaluated.
This closes the stmt buffer. The goroutine exits.
- The main goroutine tries to push some commands into the buffer only to
find that it was already closed. An assertion error is returned, and
a sentry event is created.

I think we should just not close the stmt buffer in the `errCallback`
since this was never necessary and can lead to the scenario described
above where no sentry event should be emitted.

Fixes: #79746.

Release note: None

### CODE REVIEW

**Review by cucaroach — COMMENTED** [review]
This seems fine but I don't feel qualified to sign off on it, can you get another reviewer more familiar with this code?

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80070#-:-N01JE_I38UDkwiKZTX-:b-vgmtn2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

**Comment by yuzefovich** [comment]
Unfortunately, I don't think anyone on our team has any familiarity with this code. @jordanlewis was the main reviewer of #59330 when the code in question was introduced, so maybe you can take a quick look?

**Comment by yuzefovich** [comment]
@mgartner I think you mentioned that you need to get familiar with connExecutor for UDFs, so maybe could take a look at this too?

**Comment by jordanlewis** [comment]
@yuzefovich, is there a way we can test this scenario you describe? It sounds like an interesting edge case, yet clearly it's not tested at all right now

@RichardJCai, aren't you working on internal executor soon? Could you take a look at this (and maybe talk more with Yahor about the change)?

**Review by yuzefovich — COMMENTED** [review]
Alright, I added a WIP commit with the test that follows exactly the scenario I described in the commit message. Everything behaves exactly as I described. However, without introducing several knobs I don't know how to make such a test non-flaky, and adding the knobs seems not worth it just for this edge case, so I'm inclined to not merge the test.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80070#-:-N0YU0da0KY4O618rePq:bo1k2w3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @jordanlewis)

**Review by mgartner — APPROVED** [review]
nit: fix "the new goroutine is span..." in the PR description

:lgtm: It's unfortunate that a test requires these complex knobs or significantly restructuring the code. i'm find with leaving out the test.

> Reviewed 1 of 1 files at r1, 3 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80070#-:-N0aJ5gSDNNfg-MAhgUW:bx5bapp)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[`-- commits` line 17 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80070#-N0aJNJJ69GfQwOPb-Bd):*
nit: formatting typo here?

**Review by yuzefovich — COMMENTED** [review]
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80070#-:-N0amQXc8n7UZY_FkB_h:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis and @mgartner)

**Pull request merged**
